README.md
# OHLC

![example workflow](https://github.com/alfetahe/ohlc/actions/workflows/elixir.yml/badge.svg)  [![hex.pm version](https://img.shields.io/hexpm/v/coverex.svg?style=flat)](https://hex.pm/packages/ohlc)


A library that can generate OHLC(open, high, low, close) candles from trade events.

It supports multiple timeframes including minute, hour, day and week and different configuration
options `OHLC.opts()`.

All generated dates and times are in UTC +0 timezone.
The library is using the built in module `DateTime` which handles the timezones.

OHLC library also provides helper functions when working with OHLC data.
Please look at the available modules and their functions in the documentation:
https://hexdocs.pm/ohlc

This library could be useful if you want to create your own charting engine or trading bot.

## Installation

The package can be installed by adding `ohlc` to your
list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ohlc, "~> 1.2"}
  ]
end
```

## Example usage
```elixir
iex>trades = [
...>  [price: 12.21, volume: 0.98, time: 1616439921],
...>  [price: 12.54, volume: 12.1, time: 1616439931],
...>  [price: 12.56, volume: 18.3, time: 1616439952],
...>  [price: 18.9, volume: 12, time: 1616440004],
...>  [price: 11, volume: 43.1, time: 1616440025],
...>  [price: 18.322, volume: 43.1, time: 1616440028]
...>]
...>
...>OHLC.create_candles(trades, :minute)
{
  :ok,
  %{
    candles: [
      %{
        close: 12.56,
        etime: 1616439959,
        high: 12.56,
        low: 12.21,
        open: 12.21,
        processed: true,
        stime: 1616439900,
        trades: 3,
        type: :bullish,
        volume: 31.38
      },
      %{
        close: 18.9,
        etime: 1616440019,
        high: 18.9,
        low: 18.9,
        open: 18.9,
        processed: true,
        stime: 1616439960,
        trades: 1,
        type: :bearish,
        volume: 12.0
      },
      %{
        close: 18.322,
        etime: 1616440079,
        high: 18.322,
        low: 11.0,
        open: 11.0,
        processed: true,
        stime: 1616440020,
        trades: 2,
        type: :bullish,
        volume: 86.2
      }
    ],
    pair: nil,
    timeframe: :minute
  }
}
```

CHANGELOG.md

# Change log

## v1.2.4 - 13.04.2022
### Added more unit tests.
### Updated readme file.

## v1.2.3 - 11.11.2022
### Added doctests and updated documentation.

## v1.2.2 - 30.06.2022
### Fixed
- Fixed OHLCFactory tests when validating the generated candles.

### Changed
- OHLCFactory generated candles will use the last closing price
as next candles opening price
- OHLCFactory default percetage change when generating candles is changed from 1 to 2

## v1.2.1 - 28.06.2022

### Added
- OHLCHelper test file
- Github Elixir action for CI

## v1.2.0 - 26.06.2022

### Added
- Factory module **OHLCFactory** which can be used to generate data.
- **gen_candles/3** function to generate OHLC candles.
- **gen_empty_candle/1** function to generate empty OHLC candle.
- **gen_trades/1** function to generate trades.

### Deprecated
- **OHLCHelper.gen_trades/6** use **OHLCFactory.gen_trades/1 instead**
- **OHLCHelper.generate_empty_candle/1** use **OHLCFactory.gen_empty_candle/1 instead**






test/test_data.exs
defmodule TestData do
  # Error return data items.
  def data_not_sequenced() do
    [
      [price: "15", volume: "15", time: "1616046310", side: "b"],
      [price: "15", volume: "15", time: "1615896167", side: "s"]
    ]
  end

  # Success return data items.
end


test/candloo_test.exs
defmodule CandlooTest do
  use ExUnit.Case
  doctest Candloo

  test "Trades list is not sequenced by date and must return an error." do
    case Candloo.create_candles(data_not_sequenced(), :minute) do
      {:error, _} -> assert(true)
      {:ok, _} -> assert(false)
      _ -> assert(false)
    end
  end

  test "Trades list does not contain all necessary keys and must return an error" do
    case Candloo.create_candles(error_in_keys(), :minute) do
      {:error, _} -> assert(true)
      {:ok, _} -> assert(false)
      _ -> assert(false)
    end
  end

  test "Trades list value for :side is wrong and must return an error" do
    case Candloo.create_candles(data_key_value_wrong(), :minute) do
      {:error, _} -> assert(true)
      {:ok, _} -> assert(false)
      _ -> assert(false)
    end
  end

  test "Must contain only single one minute candle" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert length(data[:candles]) === 1
  end

  test "Must contain two one minute candles" do
    {:ok, data} = Candloo.create_candles(two_one_minute_candles(), :minute)

    assert length(data[:candles]) === 2
  end

  test "Must contain two hourly candles" do
    trades = single_hourly_candle_1() ++ single_hourly_candle_2()
    {:ok, data} = Candloo.create_candles(trades, :hour, [:skip_no_trades])

    assert length(data[:candles]) === 2
  end

  test "First candles high = 167.5000" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).high === 167.5
  end

  test "First candles low = 0.3000" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).low === 0.3
  end

  test "First candles open = 15.5000" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).open === 15.0
  end

  test "First candles close = 11.1100" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).close === 11.11
  end

  test "Single daily candles test" do
    trades = single_daily_candle_1()

    {:ok, data} = Candloo.create_candles(trades, :day)

    assert(
      length(data[:candles]) === 1 and
     Enum.at(data[:candles], 0).open === Enum.at(trades, 0)[:price] and
     Enum.at(data[:candles], 0).close === Enum.at(trades, -1)[:price] and
     Enum.at(data[:candles], 0).volume === calculate_total_volume_trades(trades)
     )
  end

  def calculate_total_volume_trades(trades) do
    Enum.reduce(trades, fn trade, acc ->
      volume_formatted = Candloo.format_to_float(trade[:volume])
      if (is_float(acc)) do
        acc + volume_formatted
      else
        Candloo.format_to_float(acc[:volume]) + volume_formatted
      end
    end) |> Float.round(4)
  end

  # Success return data items.
  def single_minute_candle_data do
    [
      [price: "15", volume: "0.2", time: "1616436301", side: "s"],
      [price: "17", volume: "0.6", time: "1616436302", side: "b"],
      [price: "15", volume: "12", time: "1616436303", side: "s"],
      [price: "15", volume: "150", time: "1616436303", side: "b"],
      [price: "12", volume: "1.5", time: "1616436314", side: "s"],
      [price: "1", volume: "1.6", time: "1616436316", side: "b"],
      [price: "24", volume: "1.7", time: "1616436322", side: "s"],
      [price: "167", volume: "19", time: "1616436346", side: "b"],
      [price: "0.3", volume: "15", time: "1616436347", side: "s"],
      [price: "167.5", volume: "13", time: "1616436347", side: "b"],
      [price: "12", volume: "13", time: "1616436352", side: "s"],
      [price: "11", volume: "17", time: "1616436355", side: "b"],
      [price: "11", volume: "11", time: "1616436358", side: "s"],
      [price: "11.11", volume: "1125", time: "1616436359", side: "b"]
    ]
  end

  def two_one_minute_candles() do
    items = [
      [price: "15", volume: "0.2", time: "1616436287", side: "s"],
      [price: "17", volume: "0.6", time: "1616436299", side: "b"],
      [price: "17", volume: "0.6", time: "1616436300", side: "b"]
    ]

    items ++ single_minute_candle_data()
  end

  def single_hourly_candle_1 do
    [
      [price: "125.54", volume: "0.1", time: "1616436299", side: "s"],
      [price: "125.32", volume: "1.4", time: "1616436734", side: "b"],
      [price: "125.12", volume: "1.9", time: "1616437334", side: "s"],
      [price: "126.877", volume: "15", time: "1616437394", side: "s"],
      [price: "19.3", volume: "19.43", time: "1616438474", side: "b"],
      [price: "119", volume: "0.002", time: "1616439119", side: "b"],
      [price: "119.654", volume: "0.89", time: "1616439120", side: "s"],
      [price: "129.32", volume: "1.42", time: "1616439302", side: "s"],
      [price: "130.0", volume: "1", time: "1616439600", side: "b"]
    ]
  end

  def single_hourly_candle_2 do
    [
      [price: "12", volume: "22", time: "1616439602", side: "b"],
      [price: "12.56", volume: "18.3", time: "1616440572", side: "b"],
      [price: "18.9", volume: "12", time: "1616440692", side: "s"],
      [price: "11", volume: "43", time: "1616440759", side: "s"],
      [price: "199.3", volume: "8.93", time: "1616441583", side: "s"],
      [price: "2222", volume: "8", time: "1616441940", side: "b"],
      [price: "1234", volume: "8", time: "1616441952", side: "s"],
      [price: "44", volume: "7", time: "1616442512", side: "s"],
      [price: "98.4", volume: "0.5", time: "1616442679", side: "s"]
    ]
  end

  def single_daily_candle_1 do
    [
      [price: 125.54, volume: "0.1", time: "1616633999", side: "s"],
      [price: 125.32, volume: "1.4", time: "1616641905", side: "b"],
      [price: 125.12, volume: "1.9", time: "1616674974", side: "s"],
      [price: 126.877, volume: "15", time: "1616702514", side: "s"],
      [price: 19.3, volume: "19.43", time: "1616709599", side: "b"],
    ]
  end

  # Error return data items.
  def data_key_value_wrong() do
    [
      [price: "15", volume: "15", time: "1615896167", side: "b"],
      [price: "15", volume: "15", time: "1615896167", side: 3]
    ]
  end

  def data_not_sequenced() do
    [
      [price: "15", volume: "15", time: "1615896667", side: "s"],
      [price: "15", volume: "15", time: "1616046310", side: "b"],
      [price: "15", volume: "15", time: "1615896167", side: "s"]
    ]
  end

  def error_in_keys() do
    [
      [price: "15", volume: "15", time: "1615896167", side: "b"],
      [pricez: "15", volume: "15", time: "1615896168", side: "b"],
      [price: "15", volume: "15", timez: "1615896169", side: "b"]
    ]
  end
end


test/candloo_testx.exs
defmodule CandlooTest do
  use ExUnit.Case
  doctest Candloo

  test "Trades list is not sequenced by date and must return an error." do
    case Candloo.create_candles(data_not_sequenced(), :minute) do
      {:error, _} -> assert(true)
      {:ok, _} -> assert(false)
      _ -> assert(false)
    end
  end

  test "Trades list does not contain all necessary keys and must return an error" do
    case Candloo.create_candles(error_in_keys(), :minute) do
      {:error, _} -> assert(true)
      {:ok, _} -> assert(false)
      _ -> assert(false)
    end
  end

  test "Trades list value for :side is wrong and must return an error" do
    case Candloo.create_candles(data_key_value_wrong(), :minute) do
      {:error, _} -> assert(true)
      {:ok, _} -> assert(false)
      _ -> assert(false)
    end
  end

  test "Must contain only single one minute candle" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert length(data[:candles]) === 1
  end

  test "Must contain two one minute candles" do
    {:ok, data} = Candloo.create_candles(two_one_minute_candles(), :minute)

    assert length(data[:candles]) === 2
  end

  test "Must contain two hourly candles" do
    trades = single_hourly_candle_1() ++ single_hourly_candle_2()
    {:ok, data} = Candloo.create_candles(trades, :hour, [:skip_no_trades])

    assert length(data[:candles]) === 2
  end

  test "First candles high = 167.5" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).high === 167.5
  end

  test "First candles low = 0.3" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).low === 0.3
  end

  test "First candles open = 15.5" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).open === 15.0
  end

  test "First candles close = 11.11" do
    {:ok, data} = Candloo.create_candles(single_minute_candle_data(), :minute)

    assert Enum.at(data[:candles], 0).close === 11.11
  end

  # Success return data items.
  def single_minute_candle_data do
    [
      [price: "15", volume: "0.2", time: "1616436301", side: "s"],
      [price: "17", volume: "0.6", time: "1616436302", side: "b"],
      [price: "15", volume: "12", time: "1616436303", side: "s"],
      [price: "15", volume: "150", time: "1616436303", side: "b"],
      [price: "12", volume: "1.5", time: "1616436314", side: "s"],
      [price: "1", volume: "1.6", time: "1616436316", side: "b"],
      [price: "24", volume: "1.7", time: "1616436322", side: "s"],
      [price: "167", volume: "19", time: "1616436346", side: "b"],
      [price: "0.3", volume: "15", time: "1616436347", side: "s"],
      [price: "167.5", volume: "13", time: "1616436347", side: "b"],
      [price: "12", volume: "13", time: "1616436352", side: "s"],
      [price: "11", volume: "17", time: "1616436355", side: "b"],
      [price: "11", volume: "11", time: "1616436358", side: "s"],
      [price: "11.11", volume: "1125", time: "1616436359", side: "b"]
    ]
  end

  def two_one_minute_candles() do
    items = [
      [price: "15", volume: "0.2", time: "1616436287", side: "s"],
      [price: "17", volume: "0.6", time: "1616436299", side: "b"],
      [price: "17", volume: "0.6", time: "1616436300", side: "b"]
    ]

    items ++ single_minute_candle_data()
  end

  def single_hourly_candle_1 do
    [
      [price: "125.54", volume: "0.1", time: "1616436299", side: "s"],
      [price: "125.32", volume: "1.4", time: "1616436734", side: "b"],
      [price: "125.12", volume: "1.9", time: "1616437334", side: "s"],
      [price: "126.877", volume: "15", time: "1616437394", side: "s"],
      [price: "19.3", volume: "19.43", time: "1616438474", side: "b"],
      [price: "119", volume: "0.002", time: "1616439119", side: "b"],
      [price: "119.654", volume: "0.89", time: "1616439120", side: "s"],
      [price: "129.32", volume: "1.42", time: "1616439302", side: "s"],
      [price: "130.0", volume: "1", time: "1616439600", side: "b"]
    ]
  end

  def single_hourly_candle_2 do
    [
      [price: "12", volume: "22", time: "1616439602", side: "b"],
      [price: "12.56", volume: "18.3", time: "1616440572", side: "b"],
      [price: "18.9", volume: "12", time: "1616440692", side: "s"],
      [price: "11", volume: "43", time: "1616440759", side: "s"],
      [price: "199.3", volume: "8.93", time: "1616441583", side: "s"],
      [price: "2222", volume: "8", time: "1616441940", side: "b"],
      [price: "1234", volume: "8", time: "1616441952", side: "s"],
      [price: "44", volume: "7", time: "1616442512", side: "s"],
      [price: "98.4", volume: "0.5", time: "1616442679", side: "s"]
    ]
  end

  # Error return data items.
  def data_key_value_wrong() do
    [
      [price: "15", volume: "15", time: "1615896167", side: "b"],
      [price: "15", volume: "15", time: "1615896167", side: 3]
    ]
  end

  def data_not_sequenced() do
    [
      [price: "15", volume: "15", time: "1615896667", side: "s"],
      [price: "15", volume: "15", time: "1616046310", side: "b"],
      [price: "15", volume: "15", time: "1615896167", side: "s"]
    ]
  end

  def error_in_keys() do
    [
      [price: "15", volume: "15", time: "1615896167", side: "b"],
      [pricez: "15", volume: "15", time: "1615896168", side: "b"],
      [price: "15", volume: "15", timez: "1615896169", side: "b"]
    ]
  end
end


mix.exs
defmodule OHLC.MixProject do
  use Mix.Project

  def project do
    [
      app: :ohlc,
      version: "1.2.4",
      elixir: "~> 1.4",
      description: description(),
      package: package(),
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [coveralls: :test],
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, "~> 0.21", only: :dev, runtime: false},
      {:excoveralls, "~> 0.10", only: :test}
    ]
  end

  defp docs do
    [
      extras: ["README.md"],
      main: "readme"
    ]
  end

  defp description() do
    "Package for generating OHLC(open, high, low, close) candles from trades events."
  end

  defp package() do
    [
      # These are the default files included in the package
      files: ~w(lib .formatter.exs mix.exs README* LICENSE*),
      licenses: ["GPL-3.0"],
      links: %{"GitHub" => "https://github.com/alfetahe/ohlc"}
    ]
  end
end


lib/ohlc.ex
defmodule OHLC do
  @moduledoc """
  A library that can generate OHLC(open, high, low, close) candles from trade events.

  It supports multiple timeframes including minute, hour, day and week and different configuration
  options `t:opts/0`.

  ## Installation

  The package can be installed by adding `ohlc` to your
  list of dependencies in `mix.exs`:

  ```elixir
  def deps do
    [
      {:ohlc, "~> 1.2"}
    ]
  end
  ```

  ## Example usage
      iex>trades = [
      ...>  [price: 12.21, volume: 0.98, time: 1616439921],
      ...>  [price: 12.54, volume: 12.1, time: 1616439931],
      ...>  [price: 12.56, volume: 18.3, time: 1616439952],
      ...>  [price: 18.9, volume: 12, time: 1616440004],
      ...>  [price: 11, volume: 43.1, time: 1616440025],
      ...>  [price: 18.322, volume: 43.1, time: 1616440028]
      ...>]
      ...>
      ...>OHLC.create_candles(trades, :minute)
      {
        :ok,
        %{
          candles: [
            %{
              close: 12.56,
              etime: 1616439959,
              high: 12.56,
              low: 12.21,
              open: 12.21,
              processed: true,
              stime: 1616439900,
              trades: 3,
              type: :bullish,
              volume: 31.38
            },
            %{
              close: 18.9,
              etime: 1616440019,
              high: 18.9,
              low: 18.9,
              open: 18.9,
              processed: true,
              stime: 1616439960,
              trades: 1,
              type: :bearish,
              volume: 12.0
            },
            %{
              close: 18.322,
              etime: 1616440079,
              high: 18.322,
              low: 11.0,
              open: 11.0,
              processed: true,
              stime: 1616440020,
              trades: 2,
              type: :bullish,
              volume: 86.2
            }
          ],
          pair: nil,
          timeframe: :minute
        }
      }
  """

  import OHLCHelper

  @typedoc """
  Single trade.
  """
  @type trade :: [
          {:price, number()}
          | {:volume, number()}
          | {:time, number()}
        ]

  @typedoc """
  A list of trades.
  """
  @type trades :: [trade()]

  @typedoc """
  Single candle generated.
  """
  @type candle :: %{
          required(:open) => float(),
          required(:high) => float(),
          required(:low) => float(),
          required(:close) => float(),
          required(:volume) => float(),
          required(:trades) => integer(),
          required(:stime) => integer() | float(),
          required(:etime) => integer() | float(),
          required(:type) => :bullish | :bearish | nil,
          optional(:processed) => boolean()
        }

  @typedoc """
  A list of candles.
  """
  @type candles :: [candle()]

  @typedoc """
  Available timeframes for `create_candles/3`
  """
  @type timeframe :: :minute | :hour | :day | :week

  @typedoc """
  Available options for `create_candles/3`
  - `:forward_fill` - When set true copies the previous candles closing price
  to the next candle if no trades happened to be in between.
  Useful when you don't want to get empty time gap between the generated candles.
  - `:validate_trades` - When set true all trades are being validated before
  generating the candles to avoid errors and misinformation.
  - `:previous_candle` - Trades are appended to the previous candle if possible
  before generating the new candles. Useful if you want to update existing candle.
  - `:pair` - Adds the asset pair name to the returned outputs metadata.
  """
  @type opts :: [
          {:forward_fill, boolean()}
          | {:validate_trades, boolean()}
          | {:previous_candle, candle()}
          | {:pair, atom() | binary()}
        ]

  @doc """
  Function for generating candles from trades and timeframe provided.

  Parameters:
  - `t:trades/0` - A list containing all the trades. Trades must be
  chronologically arranged(ASC) by the timestamp field.
  - `t:timeframe/0` - Timeframe for the candles.
  - `t:opts/0` - Option values for the data proccessing.

  Returns a tuple containing the metadata for the candles and a list
  of generated candles.

  ## Example

      iex>trades = [
      ...>  [price: 0.12, volume: 542.98, time: 1668108995],
      ...>  [price: 0.14, volume: 212.1, time: 1668108998],
      ...>  [price: 0.17, volume: 532.77, time: 1668112595],
      ...>  [price: 0.21, volume: 123.8, time: 1668112623]
      ...>]
      ...>
      ...>OHLC.create_candles(trades, :hour, [pair: "BTC/EUR", validate_trades: true])
      {
        :ok,
        %{
          candles: [
            %{
              close: 0.14, etime: 1668110399,
              high: 0.14, low: 0.12,
              open: 0.12,
              processed: true,
              stime: 1668106800,
              trades: 2,
              type: :bullish,
              volume: 755.08
            },
            %{
              close: 0.21,
              etime: 1668113999,
              high: 0.21,
              low: 0.17,
              open: 0.17,
              processed: true,
              stime: 1668110400,
              trades: 2,
              type: :bullish,
              volume: 656.57
            }
          ],
          pair: "BTC/EUR",
          timeframe: :hour
        }
      }

  """
  @spec create_candles(trades(), timeframe(), opts() | nil) ::
          {:ok,
           %{
             :pair => binary() | atom(),
             :timeframe => timeframe(),
             :candles => candles()
           }}
          | {:error, atom()}
  def create_candles(trades, timeframe, opts \\ []) do
    candles =
      cond do
        opts[:previous_candle] === %{} or opts[:previous_candle] === nil ->
          [OHLCFactory.gen_empty_candle()]

        true ->
          [opts[:previous_candle]]
      end

    construct_candles(candles, trades, timeframe, opts)
  end

  @doc """
  Coverts candles to new timeframe.

  Parameters:

  `candles` - A list of candles.
  `timeframe`-  Must be higher then the existing candles timeframe.
  For example if candles were previously created using :hour timeframe
  then the provided timeframe cannot be :minute.

  Returns a tuple containing the list of candles with converted timeframe.

  ## Example

      iex>trades = [
      ...>  [price: 0.12, volume: 542.98, time: 1668108995],
      ...>  [price: 0.14, volume: 212.1, time: 1668108998],
      ...>  [price: 0.17, volume: 532.77, time: 1668112595],
      ...>  [price: 0.21, volume: 123.8, time: 1668112623]
      ...>]
      ...>
      ...>{:ok, %{candles: candles}} = OHLC.create_candles(trades, :minute)
      ...>OHLC.convert_timeframe(candles, :day)
      {
        :ok,
        [
          %{
            close: 0.21,
            etime: 1668124799,
            high: 0.21,
            low: 0.12,
            open: 0.12,
            processed: true,
            stime: 1668038400,
            trades: 4,
            type: :bullish,
            volume: 1411.65
          }
        ]
      }

  """
  @spec convert_timeframe(candles(), timeframe()) :: {:ok, candles()}
  def convert_timeframe(candles, timeframe) do
    candles = timeframe_conv_loop(candles, timeframe, [], 0)

    {:ok, candles}
  end

  @doc """
  Merges candle into another candle.

  If main candles `:stime` is 0 then fallback
  to the merge_child `:stime`.

  Parameters:
  - `main_candle` - Candle which will be merged into.
  - `child_candle` - Candle which will be merged. It is important to
  have etime less than or equal to the main candle. Meaning both candles should stay
  in the same timeframe.

  ## Example

      iex>trades1 = [
      ...>  [price: 0.12, volume: 542.98, time: 1668108995],
      ...>  [price: 0.14, volume: 212.1, time: 1668108998]
      ...>]
      ...>trades2 = [
      ...>  [price: 0.17, volume: 532.77, time: 1668112595],
      ...>  [price: 0.21, volume: 123.8, time: 1668112623]
      ...>]
      ...>{:ok, %{candles: candles1}} = OHLC.create_candles(trades1, :week)
      ...>{:ok, %{candles: candles2}} = OHLC.create_candles(trades2, :week)
      ...>OHLC.merge_child(List.first(candles1), List.first(candles2))
      {
        :ok,
        %{
          close: 0.21,
          etime: 1668383999,
          high: 0.21,
          low: 0.12,
          open: 0.12,
          processed: true,
          stime: 1667779200,
          trades: 4,
          type: :bullish,
          volume: 1411.65
        }
      }


  """
  @spec merge_child(candle(), candle()) :: {:ok, candle()} | {:error, atom()}
  def merge_child(main_candle, child_candle) do
    if main_candle[:etime] >= child_candle[:etime] do
      candle = merge_single_candle(main_candle, child_candle)

      {:ok, candle}
    else
      {:error, :unable_to_merge_child}
    end
  end

  defp timeframe_conv_loop([chd | ctl], timeframe, [cchd | cctl] = conv_candles, active_stamp) do
    conv_candles =
      if chd[:stime] >= active_stamp do
        new_candle = set_conv_candle(chd, timeframe)

        [new_candle | conv_candles]
      else
        updated_candle = merge_single_candle(cchd, chd)

        [updated_candle | cctl]
      end

    active_stamp = set_active_conv_stamp(chd[:stime], active_stamp, timeframe)

    timeframe_conv_loop(ctl, timeframe, conv_candles, active_stamp)
  end

  defp timeframe_conv_loop([chd | ctl], timeframe, [], active_stamp) do
    active_stamp = set_active_conv_stamp(chd[:stime], active_stamp, timeframe)

    new_candle = set_conv_candle(chd, timeframe)
    conv_candles = [new_candle]

    timeframe_conv_loop(ctl, timeframe, conv_candles, active_stamp)
  end

  defp timeframe_conv_loop([], _timeframe, conv_candles, _active_stamp) do
    conv_candles
  end

  defp set_active_conv_stamp(candle_stamp, active_stamp, timeframe) do
    if candle_stamp > active_stamp do
      get_time_rounded(candle_stamp, timeframe, type: :up)
    else
      active_stamp
    end
  end

  defp set_conv_candle(chd, timeframe) do
    chd
    |> Map.put(:stime, get_time_rounded(chd[:stime], timeframe, type: :down))
    |> Map.put(:etime, get_time_rounded(chd[:stime], timeframe, type: :up))
  end

  defp merge_single_candle(main_candle, merge_candle) do
    main_candle
    |> Map.update(:volume, 0.0, fn vol -> (vol + merge_candle[:volume]) |> Float.round(4) end)
    |> Map.update(:trades, 0, fn trades -> trades + merge_candle[:trades] end)
    |> Map.update(:high, 0.0, fn high ->
      if high < merge_candle[:high], do: merge_candle[:high], else: high
    end)
    |> Map.update(:low, 0.0, fn low ->
      if low < merge_candle[:low] and low !== 0.0, do: low, else: merge_candle[:low]
    end)
    |> Map.put(:close, merge_candle[:close])
    |> Map.update(:type, nil, fn _type ->
      get_candle_type(main_candle[:open], merge_candle[:close])
    end)
    |> Map.update(:open, 0.0, fn open -> if open === 0.0, do: merge_candle[:open], else: open end)
  end

  defp construct_candles(candles, trades, timeframe, opts) do
    if opts[:validate_trades] do
      case validate_data(candles, trades) do
        {:error, msg} ->
          {:error, msg}

        :ok ->
          set_return_data(candles, trades, timeframe, opts)
      end
    else
      set_return_data(candles, trades, timeframe, opts)
    end
  end

  defp set_return_data(candles, trades, timeframe, opts) do
    data = %{
      :pair => opts[:pair],
      :timeframe => timeframe,
      :candles => loop_trades(trades, candles, timeframe, opts) |> Enum.reverse()
    }

    {:ok, data}
  end

  # Loops thru trades and creates or updates candles.
  defp loop_trades(
         [trades_head | trades_tail] = trades,
         [candles_head | candles_body] = candles,
         timeframe,
         opts
       ) do
    formatted_trade_data = format_trade_data(trades_head)

    dates_match =
      dates_match_timeframe(candles_head[:etime], formatted_trade_data[:time], timeframe)

    [candles, trades_tail] =
      cond do
        # Appends new candle to the candles list without the unprocessed candle.
        !candles_head[:processed] ->
          candle = create_candle(formatted_trade_data, timeframe)
          [[candle] ++ candles_body, trades_tail]

        # Updates last candle.
        dates_match === :eq ->
          updated_candle = update_candle(candles_head, formatted_trade_data)
          [[updated_candle] ++ candles_body, trades_tail]

        # Creates new candle or candles.
        dates_match === :lt or dates_match === :gt or dates_match === :empty_first_date ->
          case opts[:forward_fill] do
            true ->
              copy_or_create_loop([candles_head | candles_body], trades, timeframe)

            _ ->
              candle = create_candle(formatted_trade_data, timeframe)
              [[candle] ++ candles, trades_tail]
          end
      end

    loop_trades(trades_tail, candles, timeframe, opts)
  end

  # Returns all candles.
  defp loop_trades(trades, candles, _timeframe, _no_trade_option) when length(trades) == 0,
    do: candles

  defp copy_or_create_loop(
         [candles_head | _candles_body] = candles,
         [trades_head | trades_tail] = trades,
         timeframe
       ) do
    trade_formatted = format_trade_data(trades_head)
    candles_head_etime_added = get_time_rounded(candles_head[:etime], timeframe, type: :jump)

    date_check =
      dates_match_timeframe(
        trade_formatted[:time],
        candles_head_etime_added,
        timeframe
      )

    cond do
      date_check === :eq or date_check === :lt ->
        candle = create_candle(trade_formatted, timeframe)
        [[candle] ++ candles, trades_tail]

      date_check === :gt ->
        copied_candle =
          forward_candle(
            candles_head[:close],
            get_time_rounded(candles_head[:etime], timeframe, type: :up),
            candles_head_etime_added
          )

        candles = [copied_candle] ++ candles

        copy_or_create_loop(candles, trades, timeframe)
    end
  end

  defp forward_candle(last_price, stime, etime) do
    OHLCFactory.gen_empty_candle()
    |> Map.put(:open, last_price)
    |> Map.put(:close, last_price)
    |> Map.put(:stime, stime)
    |> Map.put(:etime, etime)
    |> Map.put(:processed, true)
  end

  # Creates new candle.
  defp create_candle(trade, timeframe) do
    type = get_candle_type(trade[:price], trade[:price])

    OHLCFactory.gen_empty_candle()
    |> Map.put(:open, trade[:price])
    |> Map.put(:close, trade[:price])
    |> Map.put(:high, trade[:price])
    |> Map.put(:low, trade[:price])
    |> Map.put(:volume, trade[:volume])
    |> Map.put(:trades, 1)
    |> Map.put(:type, type)
    |> Map.put(:stime, get_time_rounded(trade[:time], timeframe, type: :down))
    |> Map.put(:etime, get_time_rounded(trade[:time], timeframe))
    |> Map.put(:processed, true)
  end

  # Returns updated candle.
  defp update_candle(candle, trade) do
    type = get_candle_type(candle[:open], trade[:price])

    %{
      candle
      | :close => trade[:price],
        :high => max(trade[:price], candle[:high]) |> Float.round(4),
        :low => min(trade[:price], candle[:low]) |> Float.round(4),
        :volume => (trade[:volume] + candle[:volume]) |> Float.round(4),
        :trades => 1 + candle[:trades],
        :type => type,
        :processed => true
    }
  end

  # Formats the trade data.
  defp format_trade_data(trade) do
    trade = Keyword.put(trade, :time, format_to_float(trade[:time]) |> round())
    trade = Keyword.put(trade, :price, format_to_float(trade[:price]) |> Float.round(4))
    trade = Keyword.put(trade, :volume, format_to_float(trade[:volume]) |> Float.round(4))

    trade
  end

  defp dates_match_timeframe(first_date, second_date, timeframe) do
    if first_date !== 0 do
      first_date = get_time_rounded(first_date, timeframe, format: :struct)
      second_date = get_time_rounded(second_date, timeframe, format: :struct)

      DateTime.compare(first_date, second_date)
    else
      :empty_first_date
    end
  end
end
